src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs

```diff
@@ -63,14 +64,18 @@ internal class SniOptionsSelector
 
                     if (_fallbackServerCertificateSelector is null)
                     {
-                        // Cache the fallback ServerCertificate since there's no fallback ServerCertificateSelector taking precedence. 
-                        sslOptions.ServerCertificate = fallbackHttpsOptions.ServerCertificate;
+                        // Cache the fallback ServerCertificate since there's no fallback ServerCertificateSelector taking precedence.
+                        serverCert = fallbackHttpsOptions.ServerCertificate;
+                        intermediates = fallbackHttpsOptions.ServerCertificateIntermediates;
                     }
                 }
 
-                if (!certifcateConfigLoader.IsTestMock && sslOptions.ServerCertificate is X509Certificate2 cert2)
+                if (serverCert != null)
                 {
-                    HttpsConnectionMiddleware.EnsureCertificateIsAllowedForServerAuth(cert2);
+                    // This might be do blocking IO but it'll resolve the certificate chain up front before any connections are
+                    // made to the server
+                    sslOptions.ServerCertificate = serverCert;
+                    sslOptions.ServerCertificateContext = SslStreamCertificateContext.Create(serverCert, intermediates);
```

Comment:
It might be more straightforward to just have developers call this themselves by putting a SslStreamCertificateContext on HttpsConnectionAdapterOptions and preferring it to ServerCertificate similar to the way we treat ServerCertificateSelector.

Reply:
Wherever we load the cert, we're going to be using the context. The callback happens to late. It needs to happen at startup.

Reply:
HttpsConnectionAdapterOptions.ServerCertificateContext or whatever-we-call-it would be set at startup though.

Reply:
What does that have to do with the SNI selector? Kestrel should *always* resolve this at startup.

Reply:
It's just a comparison to point out that even though having both SslStreamCertificateContext and ServerCertificate on HttpsConnectionAdapterOptions would be redundant, ServerCertificateSelector and ServerCertificate are already redundant, so this isn't a new problem if we decide to add SslStreamCertificateContext.

Reply:
I don't understand what this has to do with SNI. Exposing `SslStreamCertificateContext` on `HttpsConnectionAdapterOptions` is fine and I have no pushback but it has nothing to do with this code path...